File: app/src/main/java/com/tutv/android/ui/series/SeasonListAdapter.java
package com.tutv.android.ui.series;


import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.tutv.android.R;
import com.tutv.android.domain.Episode;
import com.tutv.android.domain.Season;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import kotlin.NotImplementedError;

import static android.view.View.*;
import static android.view.View.VISIBLE;

public class SeasonListAdapter extends RecyclerView.Adapter<SeasonListAdapter.ViewHolder> {

    private final List<Season> seasonList;

    public SeasonListAdapter() {
        this.seasonList = new ArrayList<>();
    }

    public void update(final List<Season> seasonList) {
        this.seasonList.clear();
        if(seasonList != null)
            this.seasonList.addAll(seasonList);

        notifyDataSetChanged();
    }

    @NonNull
    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.viewholder_season, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull ViewHolder holder, int position) {
        holder.setSeasonNumber(seasonList.get(position).getNumber());
        holder.setEpisodes(seasonList.get(position).getEpisodes());
        holder.setEpisodeListVisible(seasonList.get(position).isExpanded());
    }

    @Override
    public int getItemCount() {
        return seasonList.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        private TextView seasonNumberTextView;
        private RecyclerView seasonEpisodeRecyclerView;
        private EpisodeListAdapter episodeListAdapter;

        public ViewHolder(@NonNull @NotNull View itemView) {
            super(itemView);

            seasonNumberTextView = (TextView) itemView.findViewById(R.id.season_number);
            seasonEpisodeRecyclerView = (RecyclerView) itemView.findViewById(R.id.season_episode_recyclerview);

            episodeListAdapter = new EpisodeListAdapter();
            seasonEpisodeRecyclerView.setAdapter(episodeListAdapter);

            seasonNumberTextView.setOnClickListener(event -> {
                Season season = seasonList.get(getAdapterPosition());
                season.setExpanded(!season.isExpanded());
                notifyItemChanged(getAdapterPosition());
            });
        }

        public void setSeasonNumber(int seasonNumber) {
            seasonNumberTextView.setText("Season " + seasonNumber);
        }

        public void setEpisodes(List<Episode> episodesList) {
            episodeListAdapter.update(episodesList);
        }

        public void setEpisodeListVisible(boolean visible) {
            seasonEpisodeRecyclerView.setVisibility(visible ? VISIBLE : GONE);
        }

    }
}

File: app/src/main/java/com/tutv/android/ui/series/SeasonListAdapter.kt
package com.tutv.android.ui.series

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.tutv.android.R
import com.tutv.android.domain.Episode
import com.tutv.android.domain.Season
import java.util.*

class SeasonListAdapter(var seasonAndEpisodeClickedListener: SeasonAndEpisodeClickedListener?) :
        RecyclerView.Adapter<SeasonListAdapter.ViewHolder?>() {

    private val seasonList: MutableList<Season?> = ArrayList()

    fun update(seasonList: List<Season>?) {
        this.seasonList.clear()
        if (seasonList != null) this.seasonList.addAll(seasonList)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.viewholder_season, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setSeasonNumber(seasonList[position]?.number)
        holder.setEpisodes(seasonList[position]?.episodes)
        holder.setEpisodeListVisible(seasonList[position]?.isExpanded)
        if (seasonList[position] != null) {
            holder.setSeasonEpisodeClickListener {
                episode: Episode -> seasonAndEpisodeClickedListener?.onClick(seasonList[position]!!, episode)
            }
        }
    }

    override fun getItemCount(): Int {
        return seasonList.size
    }

    fun updateSeason(season: Season?) {
        for (i in seasonList.indices) {
            if (seasonList[i]?.number == season?.number) {
                seasonList[i] = season
                break
            }
        }
        notifyDataSetChanged()
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val seasonNumberTextView: TextView? = itemView.findViewById<View?>(R.id.season_number) as TextView?
        private val seasonEpisodeRecyclerView: RecyclerView? = itemView.findViewById<View?>(R.id.season_episode_recyclerview) as RecyclerView?
        private val episodeListAdapter: EpisodeListAdapter = EpisodeListAdapter()

        fun setSeasonNumber(seasonNumber: Int?) {
            seasonNumberTextView?.text = "Season $seasonNumber"
        }

        fun setEpisodes(episodesList: List<Episode>?) {
            episodeListAdapter.update(episodesList)
        }

        fun setEpisodeListVisible(visible: Boolean?) {
            seasonEpisodeRecyclerView?.visibility = if (visible == true) View.VISIBLE else View.GONE
        }

        fun setSeasonEpisodeClickListener(episodeClickListener: EpisodeClickListener?) {
            episodeListAdapter.setEpisodeClickedListener(episodeClickListener)
        }

        init {
            seasonEpisodeRecyclerView?.adapter = episodeListAdapter
            seasonNumberTextView?.setOnClickListener {
                val season = seasonList[adapterPosition]
                if (season != null) {
                    season.isExpanded = !(season.isExpanded)
                }
                notifyItemChanged(adapterPosition)
            }
        }
    }

}